src/ngraph/op/avg_pool.cpp

```diff
@@ -369,12 +371,328 @@ void op::AvgPool::generate_adjoints(autodiff::Adjoints& adjoints, const NodeVect
+                                         const Shape& pads_begin,
+                                         const Shape& pads_end,
+                                         const Shape& kernel,
+                                         bool exclude_pad)
+    : Op(check_single_output_args({delta.get_node_shared_ptr()}))
+    , m_forward_arg_shape(forward_arg_shape)
+    , m_kernel(kernel)
+    , m_strides(strides)
+    , m_pads_begin(pads_begin)
+    , m_pads_end(pads_end)
+    , m_exclude_pad(exclude_pad)
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::v1::AvgPoolBackprop::validate_and_infer_types()
+{
+    // infer_batched_forward_pooling wants CoordinateDiffs for these, while the pooling ops for
+    // now still take Shape (no negative padding).
+    CoordinateDiff pads_begin(m_pads_begin.begin(), m_pads_begin.end());
+    CoordinateDiff pads_end(m_pads_end.begin(), m_pads_end.end());
+
+    PartialShape forward_result_shape = infer_batched_pooling_forward(
+        this, m_forward_arg_shape, pads_begin, pads_end, m_kernel, m_strides, m_exclude_pad);
+
+    const PartialShape& delta_shape = get_input_partial_shape(0);
+
+    NODE_VALIDATION_CHECK(
+        this,
+        forward_result_shape.compatible(delta_shape),
+        "Inferred forward output shape does not match delta shape (inferred forward output ",
+        "shape: ",
+        forward_result_shape,
+        ", delta shape: ",
+        delta_shape,
+        ").");
+
+    set_output_type(0, get_input_element_type(0), m_forward_arg_shape);
+}
+
+const Shape& op::v1::AvgPoolBackprop::get_forward_arg_shape() const
+{
+    return m_forward_arg_shape;
+}
+
+void op::v1::AvgPoolBackprop::set_forward_arg_shape(const Shape& forward_arg_shape)
+{
+    m_forward_arg_shape = forward_arg_shape;
+}
+
+const Shape& op::v1::AvgPoolBackprop::get_kernel() const
+{
+    return m_kernel;
+}
+
+void op::v1::AvgPoolBackprop::set_kernel(const Shape& kernel)
+{
+    m_kernel = kernel;
+}
+
+const Strides& op::v1::AvgPoolBackprop::get_strides() const
+{
+    return m_strides;
+}
+
+void op::v1::AvgPoolBackprop::set_strides(const Strides& strides)
+{
+    m_strides = strides;
+}
+
+const Shape& op::v1::AvgPoolBackprop::get_pads_begin() const
+{
+    return m_pads_begin;
+}
+
+void op::v1::AvgPoolBackprop::set_pads_begin(const Shape& pads_begin)
+{
+    m_pads_begin = pads_begin;
+}
+
+const Shape& op::v1::AvgPoolBackprop::get_pads_end() const
+{
+    return m_pads_end;
+}
+
+void op::v1::AvgPoolBackprop::set_pads_end(const Shape& pads_end)
+{
+    m_pads_end = pads_end;
+}
+
+bool op::v1::AvgPoolBackprop::get_exclude_pad() const
+{
+    return m_exclude_pad;
+}
+
+void op::v1::AvgPoolBackprop::set_exclude_pad(bool exclude_pad)
+{
+    m_exclude_pad = exclude_pad;
+}
+
+shared_ptr<Node> op::v1::AvgPoolBackprop::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    AvgPoolBackprop* avpn = new AvgPoolBackprop(m_forward_arg_shape,
```

Comment:
Is there any reason of not using `make_shared` here?

Reply:
I think this is carried forward from the `v0` code. No reason AFAIK; I've noticed it a few times as I've been browsing the code but we've never gotten around to fixing it.

Reply:
Please fix it both places.